crates/project-model/src/sysroot.rs

```diff
@@ -35,6 +36,15 @@ impl ops::Index<SysrootCrate> for Sysroot {
 }
 
 impl Sysroot {
+    /// Returns sysroot directory, where `bin/`, `etc/`, `lib/`, `libexec/`
+    /// subfolder live, like:
+    /// `$HOME/.rustup/toolchains/nightly-2022-07-23-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library`
+    pub fn src_root(&self) -> &AbsPath {
```

Comment:
I think the names of this and `root` should be swapped. Also it did be nice if `path/to/sysroot/lib/rustlib/src/rust` could be derived from `path/to/sysroot` if the former is not given.

Reply:
Okay, comments fixed in https://github.com/rust-lang/rust-analyzer/pull/12858/commits/696775153dffaa701bcabe67b5fe601db862c4f4

> Also it did be nice if `path/to/sysroot/lib/rustlib/src/rust` could be derived from `path/to/sysroot` if the former is not given.

We never have a case where `root` is given but `src_root` is missing: previously, `Sysroot` could only be built when passing `src_root`. And simply joining `lib/rustlib/src/rust` is not a great idea. We already have a method for that (`discover_sysroot_src_dir`) and it respects `RUST_SRC_PATH`.

I considered making it so that building a `Sysroot` always takes `root` and an optional `src_root` (and if `src_root` is missing, we use the "discover" method), but decided against it, since:

  * There's a test case with a fake sysroot that only specifies `src_root` (it doesn't reproduce the directory structure of a real sysroot). We could fix that test case, no big deal, but...
  * We support `rust-project.json`, which up until now only allowed a `sysroot_src` field, not a `sysroot` field.
  
And going from `sysroot_src` to `sysroot` seems like a bad idea - it would involve checking that the last few path components are `["lib", "rustlib", "src", "rust]`, and removing them. I don't feel great about that.

Reply:
I think we should allow only giving the `sysroot` field and skipping `sysroot_src`.

> And simply joining lib/rustlib/src/rust is not a great idea.

Why not?

Reply:
> Why not?

Because non-cargo projects (`ProjectWorkspace::Json`) might have "libstd sources" somewhere but not a "sysroot" like we expect for cargo projects, maybe? I'm not sure.

> I think we should allow only giving the `sysroot` field and skipping `sysroot_src`.

The question is where?

For cargo projects, we already call `discover_*` functions for both `root` and `src_root`. Those respect `RUST_SRC_PATH` for example.

For `ProjectWorkspace::Json` projects, I'm not sure who sets the standard (maybe RA does?), but the `rust-project.json` format currently only has a `sysroot_src` field. Is that the case where you want to allow passing `sysroot` and "guess" `sysroot_src` ?

Reply:
> Because non-cargo projects (ProjectWorkspace::Json) might have "libstd sources" somewhere but not a "sysroot" like we expect for cargo projects, maybe? I'm not sure.

Using sysroot + lib/rustlib/src/rust as fallback for sysroot_src would work fine in that case, right?

Reply:
> > Because non-cargo projects (ProjectWorkspace::Json) might have "libstd sources" somewhere but not a "sysroot" like we expect for cargo projects, maybe? I'm not sure.
> 
> Using sysroot + lib/rustlib/src/rust as fallback for sysroot_src would work fine in that case, right?

No, they specify the `lib/rustlib/src/rust` - we can't get what we need by joining additional path segments, we need to strip some, and that feels really iffy (but we can do it if we want)

There's zero codepaths where we have `sysroot` but not `sysroot_src`.

`sysroot` is a directory with `src/`, `lib/`, `libexec/`, `etc/`, etc. `sysroot_src` is a deeper path with libstd sources (sorry about the initial confusion, my comments were swapped).

Reply:
You don't have sysroot_src if the user didn't specify one. I think the behavior should be:

if user specifies sysroot_src => use it, else use sysroot + lib/rustlib/src/rust

Or to put it another way:

sysroot_src is not specified by the user, set it to sysroot + lib/rustlib/src/rust

Reply:
> You don't have sysroot_src if the user didn't specify one.

Ah, right. But right now _no one_ is specifying `sysroot` either. That would require bazel (or other build systems that generate `rust-project.json` files) to set that field.

We should really discuss the `Cargo` / `Json` cases separately 😂

Reply:
I implemented what I _think_ you and @jyn514 asked for in https://github.com/rust-lang/rust-analyzer/pull/12858/commits/2c2520fbb48b977a805b46af79e4016a4394e719